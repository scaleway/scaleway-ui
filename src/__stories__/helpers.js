export const showAlert = (text = 'fired!') => {
  // eslint-disable-next-line no-alert
  alert(text)
}

export const log = (text = 'fired!') => {
  // eslint-disable-next-line no-console
  console.log(text)
}
