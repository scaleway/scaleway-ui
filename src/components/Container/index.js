import styled from '@emotion/styled'
import PropTypes from 'prop-types'
import React from 'react'
import BorderedBox from '../BorderedBox'
import Box from '../Box'
import Typography from '../Typography'

const StyledContainer = styled(Box)`
  margin-top: 40px;
`

const StyledBorderedBox = styled(BorderedBox)`
  padding-left: 24px;
  padding-right: 24px;
  padding-top: ${({ small }) => (small ? 16 : 24)}px;
  padding-bottom: ${({ small }) => (small ? 16 : 24)}px;
  border: 1px solid
    ${({ edition, theme }) =>
      edition ? theme.colors.primary : theme.colors.gray350};
  opacity: ${({ disabled }) => (disabled ? '0.4' : 'inherit')};
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'default')};

  > * {
    margin-top: 0;
  }
  > * + * {
    margin-top: 16px;
  }
`

const Container = ({
  title,
  subtitle,
  header,
  rightTitle,
  disabled,
  edition,
  small,
  children,
  boxStyle,
  ...props
}) => (
  <StyledContainer {...props}>
    <Box display="flex" justifyContent="space-between">
      <Box
        display="flex"
        alignItems="center"
        justifyContent="flex-start"
        mb={1}
      >
        <Typography variant="lead" my={0} mr={2}>
          {title}
        </Typography>
        {subtitle}
      </Box>
      <div>{rightTitle}</div>
    </Box>
    {header}
    <StyledBorderedBox
      css={boxStyle}
      small={small}
      edition={edition}
      disabled={disabled}
    >
      {children}
    </StyledBorderedBox>
  </StyledContainer>
)

Container.defaultProps = {
  boxStyle: undefined,
  disabled: false,
  edition: false,
  header: undefined,
  rightTitle: undefined,
  small: false,
  subtitle: undefined,
  title: undefined,
}

Container.propTypes = {
  boxStyle: PropTypes.shape({}),
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  edition: PropTypes.bool,
  header: PropTypes.node,
  rightTitle: PropTypes.node,
  small: PropTypes.bool,
  subtitle: PropTypes.string,
  title: PropTypes.string,
}

export default Container
