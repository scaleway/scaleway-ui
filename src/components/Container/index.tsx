import { Interpolation, Theme } from '@emotion/react'
import styled from '@emotion/styled'
import PropTypes from 'prop-types'
import React, { FunctionComponent, ReactNode } from 'react'
import Box from '../Box'
import FlexBox from '../FlexBox'
import Typography from '../Typography'

const StyledContainer = styled(Box)`
  margin-top: 40px;
`

const StyledTitleContainer = styled(FlexBox)`
  margin-bottom: ${({ theme }) => theme.space['1']};
`

const StyledBox = styled(Box, {
  shouldForwardProp: props => !['small', 'edition'].includes(props.toString()),
})<{
  small?: boolean
  edition?: boolean
  disabled?: boolean
}>`
  padding-left: 24px;
  padding-right: 24px;
  padding-top: ${({ small }) => (small ? 16 : 24)}px;
  padding-bottom: ${({ small }) => (small ? 16 : 24)}px;
  border: 1px solid
    ${({ edition, theme }) =>
      edition
        ? theme.colorsDeprecated.primary
        : theme.colorsDeprecated.gray350};
  opacity: ${({ disabled }) => (disabled ? '0.4' : 'inherit')};
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'default')};

  > * {
    margin-top: 0;
  }
  > * + * {
    margin-top: 16px;
  }
`

export type ContainerProps = {
  boxStyle?: Interpolation<Theme>
  children: ReactNode
  disabled?: boolean
  edition?: boolean
  header?: ReactNode
  rightTitle?: ReactNode
  small?: boolean
  subtitle?: string
  title?: string
}

const Container: FunctionComponent<ContainerProps> = ({
  title,
  subtitle,
  header,
  rightTitle,
  disabled = false,
  edition = false,
  small = false,
  children,
  boxStyle,
  ...props
}) => (
  <StyledContainer {...props}>
    <FlexBox justifyContent="space-between">
      <StyledTitleContainer>
        <Typography variant="lead" my={0} mr={2}>
          {title}
        </Typography>
        {subtitle}
      </StyledTitleContainer>
      <div>{rightTitle}</div>
    </FlexBox>
    {header}
    <StyledBox
      bordered
      css={boxStyle}
      small={small}
      edition={edition}
      disabled={disabled}
    >
      {children}
    </StyledBox>
  </StyledContainer>
)

Container.propTypes = {
  boxStyle: PropTypes.shape({}),
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  edition: PropTypes.bool,
  /**
   * Header can be a string but also a component.
   */
  header: PropTypes.node,
  /**
   * Right title can be a string but also a component, like header properties does.
   */
  rightTitle: PropTypes.node,
  small: PropTypes.bool,
  subtitle: PropTypes.string,
  title: PropTypes.string,
}

export default Container
