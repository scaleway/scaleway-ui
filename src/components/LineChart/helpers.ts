import { Theme } from '@emotion/react'
import { DatumValue, Serie } from '@nivo/line'

const parse = (data?: DatumValue | null) => {
  if (typeof data === 'number') return data
  if (typeof data === 'string') return parseFloat(data)
  if (data instanceof Date) return data.getTime()

  return 0
}

export const getMaxChartValue = (preppedData?: Serie[]): number => {
  if (!preppedData?.length) return 0

  const maximum = Math.max(
    ...preppedData.map(data =>
      data.data.reduce(
        (max, { y }) => Math.max(parse(y), max),
        Number.NEGATIVE_INFINITY,
      ),
    ),
  )

  return Math.ceil(maximum + maximum * 0.1)
}

export const getMinChartValue = (preppedData?: Serie[]): number => {
  if (!preppedData?.length) return 0

  const minimum = Math.min(
    ...preppedData.map(data =>
      data.data.reduce(
        (min, { y }) => Math.min(parse(y), min),
        Number.POSITIVE_INFINITY,
      ),
    ),
  )

  return Math.floor(minimum - minimum * 0.1)
}

export const getMin = (values: number[] = []): number =>
  values.length ? Math.min(...values) : 0

export const getMax = (values: number[] = []): number =>
  values.length ? Math.max(...values) : 0

export const getAverage = (values: number[] = []): number =>
  values.length
    ? Math.round(
        (values.reduce((average, curr) => average + curr, 0) / values.length) *
          100,
      ) / 100
    : 0

export const getCurrent = (values: number[] = []): number =>
  values.length ? values[values.length - 1] : 0

export const getSelected = (
  label: string,
  index: number,
  selected: string[] = [],
): string[] => {
  const labelIndexed = label + index.toString()
  const found = selected.indexOf(labelIndexed)

  if (found > -1) {
    selected.splice(found, 1)
  } else {
    selected.push(labelIndexed)
  }

  return selected
}

export const legendColors = (theme: Theme): string[] =>
  [
    theme.colors.chartGreen,
    theme.colors.chartPurple,
    theme.colors.gray350,
  ] as string[]

export const getLegendColor = (index: number, theme: Theme): string => {
  const colors = legendColors(theme)

  return colors[index] || colors[colors.length - 1]
}
