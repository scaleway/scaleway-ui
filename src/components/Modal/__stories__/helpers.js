export const animations = [
  'fold',
  'zoom',
  'scaleUp',
  'scaleBack',
  'sketch',
  'slideTop',
  'slideBottom',
  'slideLeft',
  'slideRight',
]

export const widths = ['xxsmall', 'xsmall', 'small', 'medium', 'large']

export const placements = [
  'center',
  'top-left',
  'top',
  'top-right',
  'right',
  'bottom-right',
  'bottom',
  'bottom-left',
  'left',
]
