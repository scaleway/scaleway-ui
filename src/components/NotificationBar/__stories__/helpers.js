export default [
  'security',
  'info',
  'success',
  'warning',
  'warning-blue',
  'warning-orange',
  'alert-orange',
  'alert-red',
]
