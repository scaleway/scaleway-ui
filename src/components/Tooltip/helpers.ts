import { RefObject } from 'react'

export type TooltipPlacement = 'top' | 'right' | 'bottom' | 'left' | 'auto'
export const ARROW_WIDTH = 6 // in px
export const SPACE = 6 // in px
export const TOTAL_USED_SPACE = ARROW_WIDTH + SPACE // in px
export const DEFAULT_POSITIONS = {
  arrowLeft: 50,
  arrowTop: 99,
  arrowTransform: 'translate(-50%, -50)',
  left: 0,
  rotate: 135,
  tooltipInitialPosition: 'translate3d(0, 0, 0))',
  top: 0,
}

export const TOOLTIP_INITIAL_POSITION = {
  auto: `translate3d(0, ${TOTAL_USED_SPACE}px, 0)`,
  bottom: `translate3d(0, -${TOTAL_USED_SPACE}px, 0)`,
  left: `translate3d(${TOTAL_USED_SPACE}px, 0, 0)`,
  right: `translate3d(-${TOTAL_USED_SPACE}px, 0, 0)`,
  top: `translate3d(0, ${TOTAL_USED_SPACE}px, 0)`,
}

type ComputePlacementTypes = {
  childrenStructuredRef: DOMRect
  tooltipStructuredRef: DOMRect
}

/**
 * This function will find the best placement in a window for tooltip based on children position and tooltip size
 */
export const computePlacement = ({
  childrenStructuredRef,
  tooltipStructuredRef,
}: ComputePlacementTypes) => {
  const {
    top: childrenTop,
    left: childrenLeft,
    right: childrenRight,
  } = childrenStructuredRef

  const { width: tooltipWidth, height: tooltipHeight } = tooltipStructuredRef

  if (childrenTop - tooltipHeight - TOTAL_USED_SPACE < 0) {
    return 'bottom'
  }

  if (childrenLeft - tooltipWidth - TOTAL_USED_SPACE < 0) {
    return 'right'
  }

  if (childrenRight + tooltipWidth + TOTAL_USED_SPACE > window.innerWidth) {
    return 'left'
  }

  return 'top'
}

type ComputePositionsTypes = {
  placement: TooltipPlacement
  childrenRef: RefObject<HTMLDivElement>
  tooltipRef: RefObject<HTMLDivElement>
}

/**
 * This function will compute the positions of tooltip and arrow based on children position and tooltip size
 */
export const computePositions = ({
  placement,
  childrenRef,
  tooltipRef,
}: ComputePositionsTypes) => {
  const childrenStructuredRef = (
    childrenRef.current as HTMLDivElement
  ).getBoundingClientRect()
  const tooltipStructuredRef = (
    tooltipRef.current as HTMLDivElement
  ).getBoundingClientRect()

  const placementBasedOnWindowSize =
    placement === 'auto'
      ? computePlacement({
          childrenStructuredRef,
          tooltipStructuredRef,
        })
      : placement

  const {
    top: childrenTop,
    left: childrenLeft,
    right: childrenRight,
    width: childrenWidth,
    height: childrenHeight,
  } = childrenStructuredRef

  const { width: tooltipWidth, height: tooltipHeight } = tooltipStructuredRef

  switch (placementBasedOnWindowSize) {
    case 'bottom':
      return {
        arrowLeft: tooltipWidth / 2,
        arrowTop: -ARROW_WIDTH - 5,
        arrowTransform: '',
        left: childrenLeft + childrenWidth / 2 - tooltipWidth / 2,
        rotate: 180,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.bottom,
        top: childrenTop + childrenHeight + ARROW_WIDTH + SPACE,
      }
    case 'left':
      return {
        arrowLeft: tooltipWidth + ARROW_WIDTH + 5,
        arrowTop: tooltipHeight / 2,
        arrowTransform: 'translate(-50%, -50%)',
        left: childrenLeft - tooltipWidth - ARROW_WIDTH - SPACE * 2,
        rotate: -90,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.left,
        top: childrenTop - tooltipHeight / 2 + childrenHeight / 2,
      }
    case 'right':
      return {
        arrowLeft: -ARROW_WIDTH - 5,
        arrowTop: tooltipHeight / 2,
        arrowTransform: 'translate(50%, -50%)',
        left: childrenRight + ARROW_WIDTH + SPACE * 2,
        rotate: 90,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.right,
        top: childrenTop - tooltipHeight / 2 + childrenHeight / 2,
      }
    default: // top placement is default value
      return {
        arrowLeft: tooltipWidth / 2,
        arrowTop: tooltipHeight - 1,
        arrowTransform: '',
        left: childrenLeft + childrenWidth / 2 - tooltipWidth / 2,
        rotate: 0,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.top,
        top: childrenTop - tooltipHeight - ARROW_WIDTH - SPACE,
      }
  }
}
