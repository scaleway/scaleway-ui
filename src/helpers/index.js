export * from './clipboard'

export const parseIntOr = (str, fallback) => {
  const value = Number.parseInt(str, 10)
  return !Number.isNaN(value) ? value : fallback
}

export const onKeyOnlyNumbers = e => {
  // `e.which` and `e.keyCode` are deprecated (https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent)
  const keyCode = e.key.charCodeAt(0)
  if (keyCode < 48 || keyCode > 57) {
    e.preventDefault()
  }
}
