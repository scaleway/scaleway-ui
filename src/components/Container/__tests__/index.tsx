import React from 'react'
import Container from '..'
import { shouldMatchEmotionSnapshot } from '../../../helpers/jestHelpers'

describe('Container', () => {
  test('renders correctly', () =>
    shouldMatchEmotionSnapshot(<Container>Hello</Container>))

  test('renders correctly on edition mode', () =>
    shouldMatchEmotionSnapshot(<Container edition>Hello</Container>))

  test('renders correctly with small variant', () =>
    shouldMatchEmotionSnapshot(<Container small>Hello</Container>))

  test('renders correctly when disabled', () =>
    shouldMatchEmotionSnapshot(<Container disabled>Hello</Container>))
})
