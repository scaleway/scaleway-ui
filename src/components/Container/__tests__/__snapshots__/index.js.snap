// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container renders correctly 1`] = `
<DocumentFragment>
  .emotion-1 {
  margin-top: 40px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-5.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  margin-bottom: 8px;
}

.emotion-8 {
  color: #4a4f62;
  font-weight: 400;
  margin-bottom: 0;
  margin-top: 0;
  color: #151a2d;
  font-size: 25px;
  line-height: 25px;
  margin-bottom: 0;
}

.emotion-8.emotion-8 {
  margin-top: 0;
  margin-bottom: 0;
  margin-right: 16px;
}

.emotion-11 {
  padding: 24px;
  border-radius: 4px;
  border: 1px solid #d4dae7;
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 24px;
  padding-bottom: 24px;
  border: 1px solid #d4dae7;
  opacity: inherit;
  cursor: default;
}

.emotion-11>* {
  margin-top: 0;
}

.emotion-11>*+* {
  margin-top: 16px;
}

<div
    class="emotion-0 emotion-1 emotion-2"
  >
    <div
      class="emotion-3 emotion-4"
    >
      <div
        class="emotion-5 emotion-2"
      >
        <h2
          class="emotion-7 emotion-8 emotion-2"
        />
      </div>
      <div />
    </div>
    <div
      class="emotion-10 emotion-11 emotion-2"
    >
      Hello
    </div>
  </div>
</DocumentFragment>
`;

exports[`Container renders correctly on edition mode 1`] = `
<DocumentFragment>
  .emotion-1 {
  margin-top: 40px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-5.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  margin-bottom: 8px;
}

.emotion-8 {
  color: #4a4f62;
  font-weight: 400;
  margin-bottom: 0;
  margin-top: 0;
  color: #151a2d;
  font-size: 25px;
  line-height: 25px;
  margin-bottom: 0;
}

.emotion-8.emotion-8 {
  margin-top: 0;
  margin-bottom: 0;
  margin-right: 16px;
}

.emotion-11 {
  padding: 24px;
  border-radius: 4px;
  border: 1px solid #d4dae7;
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 24px;
  padding-bottom: 24px;
  border: 1px solid #4f0599;
  opacity: inherit;
  cursor: default;
}

.emotion-11>* {
  margin-top: 0;
}

.emotion-11>*+* {
  margin-top: 16px;
}

<div
    class="emotion-0 emotion-1 emotion-2"
  >
    <div
      class="emotion-3 emotion-4"
    >
      <div
        class="emotion-5 emotion-2"
      >
        <h2
          class="emotion-7 emotion-8 emotion-2"
        />
      </div>
      <div />
    </div>
    <div
      class="emotion-10 emotion-11 emotion-2"
    >
      Hello
    </div>
  </div>
</DocumentFragment>
`;

exports[`Container renders correctly when disabled 1`] = `
<DocumentFragment>
  .emotion-1 {
  margin-top: 40px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-5.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  margin-bottom: 8px;
}

.emotion-8 {
  color: #4a4f62;
  font-weight: 400;
  margin-bottom: 0;
  margin-top: 0;
  color: #151a2d;
  font-size: 25px;
  line-height: 25px;
  margin-bottom: 0;
}

.emotion-8.emotion-8 {
  margin-top: 0;
  margin-bottom: 0;
  margin-right: 16px;
}

.emotion-11 {
  padding: 24px;
  border-radius: 4px;
  border: 1px solid #d4dae7;
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 24px;
  padding-bottom: 24px;
  border: 1px solid #d4dae7;
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-11>* {
  margin-top: 0;
}

.emotion-11>*+* {
  margin-top: 16px;
}

<div
    class="emotion-0 emotion-1 emotion-2"
  >
    <div
      class="emotion-3 emotion-4"
    >
      <div
        class="emotion-5 emotion-2"
      >
        <h2
          class="emotion-7 emotion-8 emotion-2"
        />
      </div>
      <div />
    </div>
    <div
      class="emotion-10 emotion-11 emotion-2"
      disabled=""
    >
      Hello
    </div>
  </div>
</DocumentFragment>
`;

exports[`Container renders correctly with small variant 1`] = `
<DocumentFragment>
  .emotion-1 {
  margin-top: 40px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-5.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  margin-bottom: 8px;
}

.emotion-8 {
  color: #4a4f62;
  font-weight: 400;
  margin-bottom: 0;
  margin-top: 0;
  color: #151a2d;
  font-size: 25px;
  line-height: 25px;
  margin-bottom: 0;
}

.emotion-8.emotion-8 {
  margin-top: 0;
  margin-bottom: 0;
  margin-right: 16px;
}

.emotion-11 {
  padding: 24px;
  border-radius: 4px;
  border: 1px solid #d4dae7;
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 16px;
  padding-bottom: 16px;
  border: 1px solid #d4dae7;
  opacity: inherit;
  cursor: default;
}

.emotion-11>* {
  margin-top: 0;
}

.emotion-11>*+* {
  margin-top: 16px;
}

<div
    class="emotion-0 emotion-1 emotion-2"
  >
    <div
      class="emotion-3 emotion-4"
    >
      <div
        class="emotion-5 emotion-2"
      >
        <h2
          class="emotion-7 emotion-8 emotion-2"
        />
      </div>
      <div />
    </div>
    <div
      class="emotion-10 emotion-11 emotion-2"
    >
      Hello
    </div>
  </div>
</DocumentFragment>
`;
